client/packages/electron/src/keyboardScanner/keyboardScanner.ts

```diff
@@ -0,0 +1,231 @@
+  input.isAutoRepeat;
+
+type KeyboardScannerSettings = {
+  maxMSBetweenScannerKeyInputs: number;
+  minBarcodeLength: number;
+};
+
+const KEYBOARD_SCANNER: BarcodeScanner = {
+  vendorId: 1,
+  productId: 1,
+  connected: true,
+};
+
+export class KeyboardScanner {
+  window: BrowserWindow;
+  interval: NodeJS.Timer | undefined;
+  // Buffer can be written concurrent in event handles and when processing input
+  // this lock is to make sure it's edited syncrhonously
+  lockBuffer: boolean;
+  // In scanning mode, when non barcode data entry is made, should allow it (no preventDefault())
+  allowEvents: boolean;
+  scanning: boolean;
+  buffer: KeyEvent[];
+  settings: KeyboardScannerSettings;
+
+  constructor(window: BrowserWindow) {
+    // this.device = this.findDevice();
+    this.window = window;
+    this.scanning = false;
+    this.lockBuffer = false;
+    this.allowEvents = false;
+    this.settings = {
+      maxMSBetweenScannerKeyInputs: 50,
+      minBarcodeLength: 5,
+    };
+    this.buffer = [];
+
+    this.bindInputEvent();
+  }
+
+  start() {
+    this.scanning = true;
+    // Process input, find sequence of input events that are less then maxMSBetweenScannerKeyInputs apart
+    // and at least minBarcodeLength long
+    this.interval = setInterval(() => {
+      this.processInputs();
+    }, PROCESSOR_INTERVAL);
+  }
+
+  stop() {
+    this.scanning = false;
+    if (this.interval) {
+      clearInterval(this.interval);
+      this.interval = undefined;
+    }
+  }
+
+  async processInputs() {
+    if (this.lockBuffer) return;
+    try {
+      this.lockBuffer = true;
+
+      let checkSequence: KeyEvent[] = [];
+      let passThrough: KeyEvent[] = [];
+      let barcode: KeyEvent[] = [];
+
+      this.buffer.forEach((current, index) => {
+        // Get ms for next input event or now
+        const nextMs = this.buffer[index + 1]?.ms ?? Date.now();
+        let diff = nextMs - current.ms;
+
+        checkSequence.push(current);
+        // Sequence is still ongoing
+        // ms diff between every event is sequence is below  maxMSBetweenScannerKeyInputs
+        if (diff < this.settings.maxMSBetweenScannerKeyInputs) return;
+        // If exceeding maxMSBetweenScannerKeyInputs and sequence length
+        // is not long enough to be barcode, pass all input through
+        if (checkSequence.length < this.settings.minBarcodeLength) {
+          passThrough = [...passThrough, ...checkSequence];
+          checkSequence = [];
+        } else {
+          // If exceeding maxMSBetweenScannerKeyInputs but sequnce length
+          // is long enought to be barcode, record as barcode;
+          // Very unlikely that two barcodes will be scanned within PROCESSOR_INTERVAL
+          barcode = checkSequence;
+          checkSequence = [];
+        }
+      });
+
+      // Remember left over for next check
+      this.buffer = checkSequence;
+      if (barcode.length != 0) {
+        console.log(JSON.stringify(barcode, null, ' '));
```

Comment:
not required?

Reply:
Hmm i kind of really want this one in, in case there is an issue can ask to open app with terminal and see what the barcode scan was, hmm, can I force console log in window (and maybe debug console log) ? Will try this

Reply:
https://github.com/openmsupply/open-msupply/compare/Use-keyboard-scanner...Log-electron-to-console, maybe could do something like this ¯\_(ツ)_/¯ I am also ok with deleting this log, although I did find it useful, that's how I figured out about the special separator character

Reply:
pushing it out - to log to the render console would be handy. it's not easy to see the electron console in production, but the web app / renderer console is ok. and sure - fine to leave in console logging that's useful / deliberate. Maybe add a comment, so that it isn't nit-picked in future? `// this console log is super useful and is supposed to be here!`

Reply:
Great idea, thanks: 4de4cfee028093782782141a22456d9381690c29